Refactor shell sessions injectables

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

~Blocked on #5399~

- This work fixes some of the types related to the communication channels
- Converts the IPC request channel for shell authentication to be injectable
- Simplify the local and node shell injectables to be functions all the way down.

Nokel81 committed:
Cleanup shell sessions

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove console.log

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Iku-turso commented on src/main/lens-proxy/lens-proxy.injectable.ts:
```diff
@@ -4,32 +4,24 @@
-  },
+  instantiate: (di) => new LensProxy({
+    router: di.inject(routerInjectable),
+    proxy: httpProxy.createProxy(),
```
Extract probable side-effect in injectable with `causesSideEffects`?

Nokel81 replied:
I have checked and it doesn't look like it has any side effects, should I still do this?

Iku-turso replied:
I think so yes. It is an external framework which currently lacks full understanding. Eg. if there's multiple places which `httpProxy.createProxy()`, will they step on each others toes? Maybe if the instances share a port or something? If there's a possibility of such, we'd do well to call it a side-effect :)

Also unit-testability would benefit from overridability of this.

Nokel81 replied:
All extract to an injectable but I won't mark it as "causesSideEffects" because it doesn't seem entirely necessary yet.

Nokel81 committed:
Make IPC usage injectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Convert shell authentication IPC to injectable RequestChannel

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix request bases tests as serialization is now down via electron, add general override in applicationBuilder for randomBytes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Iku-turso commented on src/renderer/utils/channel/send-to-main.injectable.ts:
```diff
@@ -3,21 +3,16 @@
   instantiate: (di) => {
     const ipcRenderer = di.inject(ipcRendererInjectable);
 
-    // TODO: Figure out way to improve typing in internals
```
Nice to get rid of this! :)

Iku-turso commented on src/renderer/utils/sync-box/provide-initial-values-for-sync-boxes.injectable.ts:
```diff
@@ -20,6 +20,11 @@ const provideInitialValuesForSyncBoxesInjectable = getInjectable({
+          initialValues,
+          type: typeof initialValues,
+        });
+
```
Forgotten debug-stuff?

Iku-turso commented on src/test-utils/channel-fakes/override-messaging-from-main-to-window.ts:
```diff
@@ -48,9 +47,7 @@ export const overrideMessagingFromMainToWindow = (mainDi: DiContainer) => {
           );
         }
 
-        const message = tentativeParseJson(data[0]);
```
Hmm. I wonder why `tentativeParseJson` is not required anymore. It would be important to still have serialisation and parsing of data over IPC as part of unit testing.

Nokel81 replied:
Why does it need to be unit tested? The IPC implementation itself does the serialisation and follows the structured clone algorithm

Iku-turso commented on src/common/utils/channel/message-channel-injection-token.ts:
```diff
@@ -4,9 +4,9 @@
+import type { IpcValue } from "./allowed-types";
 
-export interface MessageChannel<Message extends JsonValue | void = void> {
+export interface MessageChannel<Message extends IpcValue | void = void> {
```
Nit: New types introduced here add behaviour, making this PR something else than refactoring :)

Nokel81 replied:
Well it is more of back tracking the behaviour that was removed when these types were introduced.

Iku-turso commented on src/common/utils/channel/message-channel-injection-token.ts:
```diff
@@ -4,9 +4,9 @@
+import type { IpcValue } from "./allowed-types";
 
-export interface MessageChannel<Message extends JsonValue | void = void> {
+export interface MessageChannel<Message extends IpcValue | void = void> {
```
Unit tests for new types are missing.

Nokel81 replied:
Why do we need to unit test a dependency?

Nokel81 replied:
Discussed elsewhere

Iku-turso commented on src/common/utils/tentative-parse-json.ts:
```diff
@@ -1,15 +0,0 @@
-/**
```
Good to get rid of these.

Iku-turso commented on src/common/utils/tentative-stringify-json.ts:
```diff
@@ -1,15 +0,0 @@
-/**
```
Good to get rid of these.

Iku-turso commented on src/main/lens-proxy/proxy-functions/shell/api-request.injectable.ts:
```diff
@@ -21,7 +21,7 @@ const shellApiRequestInjectable = getInjectable({
     return ({ req, socket, head }) => {
       const cluster = getClusterForRequest(req);
-      const { searchParams } = new URL(req.url);
+      const { searchParams } = new URL(req.url, "https://127.0.0.1");
```
No behavioral unit test for this new feature?

Nokel81 replied:
This isn't a new feature, just an implementation change. `req.url` doesn't contain the `protocol` or `host` or `username` or `password` fields of a URL, so we need to provide a default so that it parses correctly. We don't care about those so just using a dummy value is fine.

Iku-turso commented on src/main/start-main-application/lens-window/application-window/application-window.injectable.ts:
```diff
@@ -50,7 +50,7 @@ const applicationWindowInjectable = getInjectable({
         });
 
-        await viewHasLoaded;
+        await Promise.race([viewHasLoaded, new Promise(resolve => setTimeout(resolve, 20_000))]);
```
No behavioral unit test for this new feature?

Iku-turso replied:
This alarmingly odd code should explain why it is required.

Nokel81 replied:
oh sorry, I should have removed this. This was part of testing because I couldn't get the window to open.

Iku-turso commented on src/main/utils/channel/channel-listeners/enlist-message-channel-listener.injectable.ts:
```diff
@@ -16,13 +15,7 @@ const enlistMessageChannelListenerInjectable = getInjectable({
-          handler,
-        );
-      };
+      const nativeOnCallback = (_: IpcMainEvent, message: unknown) => toJS(handler(message));
```
Nit: could use `pipeline`for left-to-right, top-to-bottom -readability.

```
pipeline(message, handler, toJS)
```

Nokel81 replied:
For function calling, I prefer the reverse-Polish notation style that comes with nested brackets.

Iku-turso commented on src/main/utils/channel/message-to-channel.injectable.ts:
```diff
@@ -3,34 +3,25 @@
-        lensWindow.send({ channel: channel.id, data: stringifiedMessage ? [stringifiedMessage] : [] }),
-      );
-    };
+        window.send({ channel: channel.id, data: message !== undefined ? [message] : [] });
```
Unit test missing for `message === undefined`?

Nokel81 replied:
As talked about elsewhere electron does the serialisation and thus we probably can also just simplify this as well.

Iku-turso commented on src/renderer/api/terminal-api.ts:
```diff
@@ -94,6 +94,8 @@ export class TerminalApi extends WebSocketApi<TerminalEvents> {
       tabId: this.query.id,
     });
+
+    console.log("authTokenArray", authTokenArray);
```
Forgotten debug-stuff :)

Iku-turso commented on src/common/utils/channel/enlist-request-channel-listener-injection-token.ts:
```diff
@@ -4,11 +4,11 @@
 export type EnlistRequestChannelListener = <
   TChannel extends RequestChannel<any, any>,
->(listener: RequestChannelListener<TChannel>) => () => void;
+>(listener: RequestChannelHandlerDescriptor<TChannel>) => () => void;
```
Nit: `descriptor` is a loaded term which already has a purpose in JavaScript, making this a tad confusing.

Iku-turso commented on src/common/utils/channel/listening-of-channels.injectable.ts:
```diff
@@ -17,7 +17,7 @@ const listeningOfChannelsInjectable = getInjectable({
     const enlistRequestChannelListener = di.inject(enlistRequestChannelListenerInjectionToken);
     const messageChannelListeners = di.injectMany(messageChannelListenerInjectionToken);
-    const requestChannelListeners = di.injectMany(requestChannelListenerInjectionToken);
+    const requestChannelListeners = di.injectMany(requestChannelHandlerInjectionToken);
```
Nit: changed naming of injectable is not propagated to variables.

Iku-turso commented on src/common/utils/channel/request-channel-injection-token.ts:
```diff
@@ -4,17 +4,19 @@
-  id: string;
-  _requestSignature?: Request;
-  _responseSignature?: Response;
+  id: Request | Response extends boolean ? string : string;
```
Shouldn't `id` be always a string? I don't understand what is being attempted here :)

Nokel81 replied:
It is, but I realise that I should have just added some ignore lint lines instead. Will add some descriptive comments.

Iku-turso commented on src/main/app-paths/app-paths-request-channel-listener.injectable.ts:
```diff
@@ -3,16 +3,16 @@
   id: "app-paths-request-channel-listener",
 
-  instantiate: (di): RequestChannelListener<AppPathsChannel> => {
+  instantiate: (di): RequestChannelHandlerDescriptor<AppPathsChannel> => {
```
Nit: The naming is not symmetrical between the two channel types, as there is no `MessageChannelListenerDescriptor` to contrast `RequestChannelHandlerDescriptor`.

Iku-turso commented on src/main/shell-session/node-shell-session/open-node-shell-session.injectable.ts:
```diff
@@ -0,0 +1,27 @@
+const openNodeShellSessionInjectable = getInjectable({
+  id: "node-shell-session",
+
+  instantiate: (): OpenNodeShellSession => (args) => new NodeShellSession(args).open(),
```
`new NodeShellSession(args).open()`
makes this `causesSideEffects: true`

Iku-turso commented on src/renderer/components/test-utils/get-application-builder.tsx:
```diff
@@ -120,6 +121,21 @@ export const getApplicationBuilder = () => {
   rendererDi.override(storesAndApisCanBeCreatedInjectable, () => true);
   mainDi.override(clusterStoreInjectable, () => clusterStoreStub);
 
+  mainDi.override(randomBytesInjectable, () => {
```
Override has side-effect. Could this override return a static value instead? The answer could be no if some recurring code that is uninteresting in the general use-case actually requires differing return values in order to not blow up.

Nokel81 replied:
Will try.

Nokel81 replied:
I think this can probably be static (for the time being at least) since the only use case is authentication.

Iku-turso commented on src/renderer/utils/channel/request-from-channel.test.ts:
```diff
@@ -57,64 +53,45 @@ describe("request from channel in renderer", () => {
 
-    expect(invokeMock).toHaveBeenCalledWith(
-      "some-channel-id",
-      JSON.stringify({ some: "object" }),
```
Hmm. I wonder if the serialisation and parsing is still being unit-tested.

Nokel81 replied:
No because it is completely handled in our dependency.

Iku-turso reviewed (changes requested):
Love the typing stuff! Main gripes: lack of unit testing, uncertainty in how serialisation and parsing is tested for IPC (I hope I'm wrong about this).

Nokel81 committed:
Respond to PR comments

- Add helper function to improve types with MessageChannelListeners

- Remove some debug printing

- Improve some typing about application update status

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove side-effect in override

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Simply implementation of sendToChannelInLensWindow

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Cleaning up some IPC types

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix type error

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix main's navigate to route

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Switch to Writable from Mutable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move proxy server to injectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Verify messages are serializable within tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add unit testing requirement to have data serializable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix type error

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Revert 'fix type error'

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Another attempt to fix type errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix type issue when generating extension types

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove unnecessary any cast

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
make more injectable for testing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Make shell sessions even more injectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add more injectables

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Switch back to jest-canvas-mock

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jansav commented on src/main/getDiForUnitTesting.ts:
```diff
@@ -124,6 +129,9 @@ export function getDiForUnitTesting(opts: { doGeneralOverrides?: boolean } = {})
   di.preventSideEffects();
 
   if (doGeneralOverrides) {
+    di.override(isProductionInjectable, () => false);
+    di.override(isDevelopmentInjectable, () => false);
+    di.override(isTestEnvInjectable, () => true);
```
I'd say production is the most important. However, these shouldn't exist at all since there shouldn't be differences in our operational code in environments.

Nokel81 replied:
I agree about `isTestEnv`, will create a separate PR to remove it.

jansav commented on src/main/start-main-application/lens-window/application-window/send-to-channel-in-electron-browser-window.injectable.ts:
```diff
@@ -13,16 +14,16 @@ const sendToChannelInElectronBrowserWindowInjectable = getInjectable({
             [frameInfo.processId, frameInfo.frameId],
             channel,
-            ...data,
+            toJS(data),
```
This doesn't belong here. MobX -> toJS conversion should be done in usage.

Nokel81 replied:
Why, I would have thought that doing it here helps with it being more resilient and reusable.

jansav commented on src/renderer/components/test-utils/get-application-builder.tsx:
```diff
@@ -65,16 +66,16 @@ type EnableExtensions<T> = (...extensions: T[]) => void;
 
 export interface ApplicationBuilder {
-  dis: DiContainers;
+  readonly dis: DiContainers;
```
Why are you adding these here? TBH, I don't see the benefit for these, but of course we still can add them, but please not in this kind of big PR.

Nokel81 replied:
Will revert

jansav commented on src/techincal/shell-sessions/local.test.ts:
```diff
@@ -0,0 +1,41 @@
+ * Copyright (c) OpenLens Authors. All rights reserved.
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
```
typo in directory name. `techincal` -> `technical`
Same goes for name of test.

jansav commented on src/techincal/shell-sessions/local.test.ts:
```diff
@@ -0,0 +1,41 @@
+  it("renders", () => {
+    expect(result.baseElement).toMatchSnapshot();
+  });
+
```
Feels like incomplete test. And I'd like to see this as behaviour, instead of something new called `technical`. 

Name could be something like `opening-shell-terminal-in-cluster`

Nokel81 replied:
Sure I can move it to `behavioural` but I haven't figured out how to test the `canvas` rendering method from `xtermjs`
